src/AspNet.Security.OAuth.Yammer/YammerAuthenticationHandler.cs

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+using AspNet.Security.OAuth.Extensions;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http.Authentication;
+using Newtonsoft.Json.Linq;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Authentication.Yammer {
+    public class YammerAuthenticationHandler : OAuthHandler<YammerAuthenticationOptions> {
+        public YammerAuthenticationHandler(HttpClient httpClient)
+            : base(httpClient) {
+        }
+
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(string code, string redirectUri) {
+            /*
+             https://developer.yammer.com/docs/oauth-2
+             Override this method because Yamer API returns unusual response for TokenEndpoint request
+             */
+            var tokenRequestParameters = new Dictionary<string, string>()
+            {
+                { "client_id", Options.ClientId },
+                { "redirect_uri", redirectUri },
+                { "client_secret", Options.ClientSecret },
+                { "code", code },
+                { "grant_type", "authorization_code" },
+            };
+
+            var requestContent = new FormUrlEncodedContent(tokenRequestParameters);
+
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+            requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            requestMessage.Content = requestContent;
+            var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
+            if (response.IsSuccessStatusCode) {
+                var accessTokenObject = JObject.Parse(await response.Content.ReadAsStringAsync())["access_token"].Value<JObject>();
+
+                JObject payload = JObject.FromObject(new {
```

Comment:
Hum, this means that all the properties returned by Yammer's token endpoint won't be flowed. Maybe we could simply add or replace the `access_token` property to `accessTokenObject` to ensure the OAuth2 middleware gets the standard property?

Alternatively, you could add an `original_response` parameter to ensure it's not lost when creating a new `JObject`.